src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,204 @@
+                    transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+                    var lockId = Int32.Parse(@lock.Id);
+                    var record = repository.Get(lockId);
+
+                    if (record == null)
+                        throw new ObjectDisposedException("@lock", "No lock record could be found for the specified lock to be released.");
+
+                    if (record.Count <= 0)
+                        throw new ObjectDisposedException("@lock", "The specified lock has already been released.");
+
+                    record.Count--;
+                }
+                catch (Exception ex) {
+                    if (ex.IsFatal()) throw;
+                    Logger.Error(ex, "An non-fatal error occurred while trying to dispose a distributed lock with name '{0}' and ID {1}.", @lock.Name, @lock.Id);
+                }
+                finally {
+                    childLifetimeScope.Dispose();
+                }
+            }
+        }
+
+        private bool TryAcquireLock(string name, TimeSpan maxValidFor, TimeSpan? timeout, string machineName, int? threadId, out DistributedLock @lock) {
+            @lock = AcquireLockInternal(name, maxValidFor, machineName, threadId, timeout.GetValueOrDefault());
+            if (@lock != null)
+                return true;
+
+            Logger.Debug("Could not acquire a lock named '{0}'.", name);
+            return false;
+        }
+
+        private DistributedLock AcquireLock(string name, TimeSpan maxValidFor, TimeSpan? timeout, string machineName, int? threadId) {
+            var @lock = AcquireLockInternal(name, maxValidFor, machineName, threadId, timeout);
+            if (@lock != null)
+                return @lock;
+
+            throw new TimeoutException("Could not acquire a lock within the specified amount of time.");
+        }
+
+        private DistributedLock AcquireLockInternal(string name, TimeSpan maxValidFor, string machineName, int? threadId, TimeSpan? timeout = null) {
+            try {
+                DistributedLockRecord record = null;
+                var acquired = Poll(() => (record = AcquireLockRecord(name, maxValidFor, machineName, threadId)) != null, timeout);
+
+                if (acquired) {
+                    Logger.Debug("Successfully acquired a lock named '{0}'.", name);
+                    return threadId != null
+                        ? DistributedLock.ForThread(this, name, machineName, threadId.Value, record.Id.ToString())
+                        : DistributedLock.ForMachine(this, name, machineName, record.Id.ToString());
+                }
+            }
+            catch (Exception ex) {
+                Logger.Error(ex, "Error while trying to acquire a lock named '{0}'.", name);
+                throw;
+            }
+
+            return null;
+        }
+
+        private DistributedLockRecord AcquireLockRecord(string name, TimeSpan maxValidFor, string machineName, int? threadId) {
```

Comment:
Same here, but I don't like an overloaded version either since this one does something different - it is focused on retrieving, creating and updating the lock record. So it (subjectively) made more sense to use the word `Record`. I also considered naming it something like `GetOrCreateOrUpdateLockRecord`, but that looks even worse. I was also considering a potential future update where this method becomes protected virtual for sub classes to use a different means of storage (maybe some shared cache or file system), so in that light the word 'Internal' didn't seem right either. That being said, in such scenarios the 'Record' makes even less sense since sub classes wouldn't even be bothered with DB related entities.